### ISSUE TITLE
[feature request] A solution for the Next's routing system to allow dynamic routes.

### ISSUE THREAD

**Issue by httpiago** [zeit/next.js#7577]
# Feature request

## Is your feature request related to a problem? Please describe.

Next is my standard choice for creating SPA projects in React, but after using it a lot, I've come to the conclusion that its routing system is very poor, you will have a headache if you want custom and dynamic routes, for example: `/user/webpack`, `/user/dan_abramov`, `/photo/12345678/likes` ...

There is [next-routes](https://github.com/fridays/next-routes) that solves the problem but I have already used it and I was not satisfied.

## Describe the solution you'd like

Recently I discovered [a package](https://github.com/remoteinterview/zero) that also works with SSR and it has an interesting behavior that would solve this problem of Next having a poor routing system. In the words described in the documentation:

> "... **all requests sent to a route that doesn't exist are passed on to the closest parent function.**
> So if you visit `/user/dan` and there is no `/user/dan.js` but there is `/user.js`. Zero will send the request to `/user` and set req.params to `['dan']`."

With this behavior, Next can inject these parameters into getInitialProps the same way it does with query:

```js
// pages/user.js
class UserPage extends React.Component {
  static async getInitialProps({ params }) {
    const username = params[0]; // 'sophie' when user visits /user/sophie
    const userInfos = await fetch(`/api/user/${username}`).then(res => res.json());

    return { userInfos };
  }
  // ...
}
```

[View the Zero documentation](https://github.com/remoteinterview/zero#route-rewrites),

## Additional context

I just wanted to suggest this improvement because I do not know if you still intend to ["integrate" the react-router-dom with Next](https://twitter.com/rauchg/status/948318111828099072) and I think this solution will be easier as well.

**Comment by Timer** [comment]
Closing as a dupe of https://github.com/zeit/next.js/issues/4989.

However, we've implemented an experimental version you can opt-into and try: https://github.com/zeit/next.js/pull/7432.

tl;dr:
```js
// next.config.js
module.exports = {
  experimental: { dynamicRouting: true }
}
```

Create pages that are named like so:
```
pages/user/$userId.js
```
